### PR TITLE
refactor(tile-converter): refactor creation of Attribute info

### PR DESCRIPTION
**Pull request by mspivak-actionengine**

### CODE REVIEW

File: modules/i3s/src/types.ts [hunk 1]
```diff
@@ -590,7 +590,23 @@ type TextureImage = {
   length?: number[];
 };
 
-export type Attribute = 'OBJECTID' | 'string' | 'double' | 'Int32' | string;
+export enum AttributeType {
```
Comment (ibgreen) [inline modules/i3s/src/types.ts:593]: I strongly prefer new style string constants to old-school enums. This seems like a step backwards to me.

File: modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  AttributeStorageInfo,
+  ESRIField,
+  Field,
+  FieldInfo,
+  PopupInfo
+} from '@loaders.gl/i3s';
+
+import {AttributeType} from '@loaders.gl/i3s';
+
+export default class AttributeMetadataInfo {
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts:12]: Nit: Prefer that we avoid export `default`. Named exports allow the transpiler to keep track of names.

File: modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+  private createFieldAttribute(key: string, fieldAttributeType: ESRIField): Field {
+    return {
+      name: key,
+      type: fieldAttributeType,
+      alias: key
+    };
+  }
+
+  /**
+   * Generates popup info to show metadata on the map.
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts:211]: I believe in strict separation between loading / logic and presentation layers. talking about "popup" (which is very much a presentation layer concept) here in the load doesn't sit right with me. 

Either rename this to something non-UI-related, or maybe move such functionality to the application

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 4]
```diff
@@ -1174,65 +1174,22 @@ export default class I3SConverter {
...
-        const attributeType = attributeTypes[key];
-
-        const storageAttribute = createdStorageAttribute(attributeIndex, key, attributeType);
-        const fieldAttributeType = getFieldAttributeType(attributeType);
-        const fieldAttribute = createFieldAttribute(key, fieldAttributeType);
+      // Add new storage attributes, fields and create popupInfo
+      this.attributeMetadataInfo.addMetadataInfo(attributeTypesMap);
+      this.layers0!.attributeStorageInfo = this.attributeMetadataInfo.attributeStorageInfo;
+      this.layers0!.fields = this.attributeMetadataInfo.fields;
+      this.layers0!.popupInfo = this.attributeMetadataInfo.popupInfo;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:1191]: this.layers0?.attributeStorageInfo is being used in _convertResources method that is in turn being used in _createNode.
So, we can't just move the re-set out of "traverseDatasetWith".
There could be done if we implemented SceneLayer3D as a class, moved attributeMetadataInfo instance into it, and create a getter "attributeStorageInfo" in the I3SConverter class.
In this case every reading (get) attributeStorageInfo in I3SConverter will read it from AttributeMetadataInfo class directly.
Does it make sense?

File: modules/i3s/src/types.ts (hunk 1 above)
Comment (mspivak-actionengine) [reply modules/i3s/src/types.ts]: Done

File: modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts (hunk 2 above)
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts]: Done

File: modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts [hunk 5]
```diff
@@ -0,0 +1,246 @@
+import type {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/attribute-metadata-info.ts]: Done

File: modules/tile-converter/test/i3s-converter/attribute-metadata-info.spec.ts [hunk 6]
```diff
@@ -0,0 +1,575 @@
+import test from 'tape-promise/tape';
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/test/i3s-converter/attribute-metadata-info.spec.ts]: Done

File: modules/i3s/src/types.ts [hunk 7]
```diff
@@ -590,7 +590,23 @@ type TextureImage = {
...
+  /** Integer data type name for feature attributes */
+  SHORT_INT_TYPE = 'Int32'
+}
+
+type StringValues<T> = {
+  [K in keyof T]: T[K] extends string ? T[K] : never;
+}[keyof T];
+type EnumAsUnion<T> = `${StringValues<T>}`;
+
+export type Attribute = EnumAsUnion<typeof AttributeType> | string;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/i3s/src/types.ts]: Done

File: modules/i3s/src/types.ts [hunk 8]
```diff
@@ -590,7 +590,23 @@ type TextureImage = {
...
+  /** Double data type name for feature attributes */
+  DOUBLE_TYPE = 'double',
+  /** Integer data type name for feature attributes */
+  SHORT_INT_TYPE = 'Int32'
+}
+
+type StringValues<T> = {
+  [K in keyof T]: T[K] extends string ? T[K] : never;
+}[keyof T];
+type EnumAsUnion<T> = `${StringValues<T>}`;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/i3s/src/types.ts]: Removed

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done.
this.layers0!.attributeStorageInfo
changed to
this.attributeMetadataInfo.attributeStorageInfo

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 9]
```diff
@@ -450,6 +450,12 @@ export default class I3SConverter {
       this.options.maxDepth
     );
 
+    this.layers0!.attributeStorageInfo = this.attributeMetadataInfo.attributeStorageInfo;
+    this.layers0!.fields = this.attributeMetadataInfo.fields;
+    this.layers0!.popupInfo = this.attributeMetadataInfo.popupInfo;
+
+    this.layers0!.layerType = _3D_OBJECT_LAYER_TYPE;
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/i3s-converter.ts:457]: It seems we never create "Integrated mesh". The idea was when we don't have feature segmentation, we create "Integrated mesh" layer type. Now it always "3D objects".
Test it with "Frankfurt" tileset, for example

File: modules/tile-converter/test/i3s-converter/helpers/geometry-converter.spec.ts [hunk 10]
```diff
@@ -592,6 +592,84 @@ test('tile-converter(i3s)#convertB3dmToI3sGeometry - array of UINTxx should be c
   t.end();
 });
 
+test('tile-converter(i3s)#convertB3dmToI3sGeometry - should convert 64-bit attributes to strings', async (t) => {
```
Comment (belom88) [inline modules/tile-converter/test/i3s-converter/helpers/geometry-converter.spec.ts:595]: Looks like this test is from another PR...

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts (hunk 9 above)
Comment (mspivak-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done. All these set operations will be performed if there is at least one attribute.

File: modules/tile-converter/test/i3s-converter/helpers/geometry-converter.spec.ts (hunk 10 above)
Comment (mspivak-actionengine) [reply modules/tile-converter/test/i3s-converter/helpers/geometry-converter.spec.ts]: No idea why this test is shown as a new one. It has been recently added to master. It looks like this review was done before the merge of that PR. Now it looks good.

**Review by belom88 — APPROVED** [review]

**Pull request merged**